mypy/fswatcher.py

```diff
@@ -85,3 +80,24 @@ def find_changed(self) -> Set[str]:
                         # Changed file.
                         changed.add(path)
         return changed
+
+    def find_changed(self) -> AbstractSet[str]:
+        """Return paths that have changes since the last call, in the watched set."""
+        return self._find_changed(self._paths)
+
+    def update_changed(self,
+                       add: List[str],
+                       remove: List[str],
+                       update: List[str],
+                       ) -> AbstractSet[str]:
+        """Alternative to find_changed() given explicit changes.
+
+        This only calls self.fs.stat() on added or updated files, not
+        on all files.  It believes all other files are unchanged!
+
+        Implies add_watched_paths() for add and update, and
+        remove_watched_paths() for remove.
+        """
+        self.remove_watched_paths(remove)
+        self.add_watched_paths(add)
+        return self._find_changed(add + update)
```

Comment:
I didn't do a careful review, but generally this looks good. I just have one question to clarify (and it is probably worth mentioning in the docstring): Is it OK to pass files to `--remove` that were never in the build?

My motivation is to avoid duplication of work in the daemon and in the runner script that filters sources (what we do internally). If only files that were previously in the build are allowed in `--remove`, then we will need to keep full list of the checked files also in the runner script (because there is no way to check whether already deleted file should be filtered or not, if the filtering is content based, like presence of type comments/annotations).

I would prefer therefore to allow passing arbitrary files to `--remove` and explicitly document this.

Reply:
To clarify, from the current implementation in this PR it looks like it should work, I just would like to make this explicit.